File: exercise/scala/day24/src/test/scala/unit/CreateToySpec.scala
package unit

import Time.*
import com.github.javafaker.Faker
import domain.Toy
import org.scalacheck.Prop.forAll
import org.scalacheck.{Gen, Properties}
import org.scalatest.EitherValues

object CreateToySpec extends Properties("Create Toy") with EitherValues {
  val maxToys = 1_000_000_000
  val invalidStock: Gen[Int] = Gen.chooseNum(-maxToys, -1)
  val validStock: Gen[Int] = Gen.chooseNum(0, maxToys)
  val faker: Faker = Faker()

  property("can not create toy with invalid stock (<0)") =
    forAll(invalidStock) { stock =>
      Toy.create(timeProvider, faker.funnyName().name(), stock).isLeft
    }

  property("can create toy with valid stock (>= 0)") =
    forAll(validStock) { stock =>
      Toy.create(timeProvider, faker.commerce().productName(), stock).isRight
    }
}

File: exercise/ts/day24/tests/unit/CreateToy.spec.ts
import * as fc from 'fast-check';
import {Toy} from "../../src/domain/Toy";
import {Time} from "../Time";
import * as E from "fp-ts/Either";
import {faker} from '@faker-js/faker';

describe('CreateToyTests', () => {
    it('can not create toy with invalid stock (<0)', () => {
        fc.assert(
            fc.property(fc.integer().filter(x => x < 0), (stock) => {
                return E.isLeft(
                    Toy.create(Time.provider, faker.commerce.product(), stock)
                );
            })
        );
    });

    it('can create toy with valid stock (>= 0)', () => {
        fc.assert(
            fc.property(fc.integer().filter(x => x >= 0), (stock) => {
                return E.isRight(
                    Toy.create(Time.provider, faker.commerce.product(), stock)
                )
            })
        );
    });
});
